### DISCUSSION TITLE
Information Extraction from PDF

### DISCUSSION THREAD

**Discussion by Apurv3377** [pymupdf/PyMuPDF#1420] [Ideas]
Currently, PyMuPDF only supports layout extraction from the PDF. Moreover, hence does not extract different semantic structures (Table, Section, Metadata, References, List, Header, Footer) separately. 

Incorporating state-of-the-art pre-trained models can be used to improve the information extraction from PDF.
Is there any plan to support the extraction of such semantic structures in the future?

**Comment by JorjMcKie** [comment]
>  Is there any plan to support the extraction of such semantic structures in the future?

No, there is not.
The main reason is, that for most of these categories there is no physical basis in the PDF. For example, what one may view as a title or table, inside the PDF page all of these are just text, or maybe even just single, individual characters.
For PDF table detection there exist separate tools which to my knowledge all are incomplete - only deliver point solutions.

Already (Py-) MuPDF's collecting and structuring characters into spans, lines, blocks or words is a deliberate application of heuristics ... in the hope that it will make sense. But often enough, even this seemingly so obvious approach fails!

If you give me a PDF page of text, I can construct one for you with identical appearance (in a PDF viewer), which refuses however to deliver any meaningful extracted text - **_except_** if you re-arrange each individual character based on its geometrical position. Look at example files  [textmaker.pdf](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/text-extraction/textmaker.pdf) and [textmaker2.pdf](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/text-extraction/textmaker2.pdf). Try to copy / paste text from the second one with any viewer and you will see what I mean.

So bottom line is, I won't go beyond what I did with layout preservation.
But any contribution from your side in that area will certainly be very welcome.
